### PR TITLE
added better workflow for submitting changes via PR

### PR DESCRIPTION
**Pull request by rikutakei**
Let me explain a little bit about this one. 
Since this is just an opinion, you can ignore this PR if you don't think it's gonna work for the team.

There's been a lot of pull requests, reviews, additional editing/committing, and merging quite recently.

I love the activity, and the feel of getting things done and working together is amazing, but one thing that annoys me (sort of) is the constant merging and the merge commits that come with it.

Even if you want to add the smallest of the changes, you have to `merge` the upstream commits before you push your small changes to the repo. As a result, there is a commit message for your small change, *as well as the commit message for the merging that had to be done to add that change*.

I think this is unnecessary and messy - I feel like the `merge` commit messages should be reserved for submitted PR, rather than for updating your local repo.

Now, I had a quick look around for a [solution](http://kernowsoul.com/blog/2012/06/20/4-ways-to-avoid-merge-commits-in-git/) ([another relevant reference](https://codeinthehole.com/tips/pull-requests-and-other-good-practices-for-teams-using-github/)), and I think it'll work for us quite nicely.

1. Always work/edit on a separate branch locally 
1. When pulling from the `swcarpentry-ja` repo and/or your repo, use `git pull --rebase` instead to prevent unnecessary merge commit(s)
1. Once ready for PR, `push` the branch to your remote and submit a pull request from that branch to `swcarpentry-ja`

Few advantages by using this approach:
* You won't mess up **your** repo by working on a separate branch
* Able to keep working on other files without adding unrelated commit(s) to the PR *(I'm especially guilty of this one from #28, as I was using my main branch for it)*
* There is minimal chance for you to have to `merge` when pulling, as `rebase` will allow you to add your changes *on top of the existing changes*

I've edited the quickstart guide to reflect this workflow - let me know of what you think of it

### CODE REVIEW

File: quickstart.md
Comment (joelnitta) [inline quickstart.md]:
```suggestion
git remote add upstream git@github.com:swcarpentry-ja/i18n.git
```

File: quickstart.md
Comment (joelnitta) [inline quickstart.md]:
```suggestion
git pull --rebase upstream ja
```

**Review by joelnitta — CHANGES_REQUESTED** [review]
I noticed the same thing when I `pull`ed without `--rebase`. I like this approach.

One point I find a little confusing though (and this would also need to be addressed in the main README) is the use of `JA` for both the remote to fetch/pull from and its branch. What about the more typical `upstream` for the remote, and `JA` for its branch?

File: quickstart.md
Comment (TomKellyGenetics) [inline quickstart.md]: I'm not a fan of excessive merge commits either but I don't know if we can really avoid them. I don't recommend people who are inexperienced with git to be handling branches and rebase. If we push rebased commits to a remote repo, it could be a problem for someone else.

I encountered a similar issue when resolving conflicts for PR https://github.com/swcarpentry-ja/i18n/pull/32 from @rkkmk. I originally made a few branch, pulled the user repo and submitted changes as another PR. Then I realised that this wasn't necessary. It is possible to pull commits from a user repo (their branch) and merge locally. Pushing to the organisation repo merges the PR via the command-line. I recommend this workflow to merge PRs. This way, only admins are handling branches and merges where things get more difficult with Git.

It should be possible to submit a PR without doing this. I want to lower the barrier for entry as much as possible. Creating a fork or pulling to a local repo creates a different branch specific that machine/account already. It should be possible to merge. I am willing to merge PRs and resolve conflicts where needed. However it would help if member submitting pull requests would pull from the organisation repo and merge changes _before_ pushing to their account and submitting a pull request. They should also be pulling from the organisation repo _regularly_ to ensure that they're working on the latest version (to reduce conflicts with things others have already changed).

**Comment by TomKellyGenetics** [comment]
I'm not a fan of excessive merge commits either but I don't know if we can really avoid them. I don't recommend people who are inexperienced with git to be handling branches and rebase. If we push rebased commits to a remote repo, it could be a problem for someone else.

I encountered a similar issue when resolving conflicts for PR #32 from @rkkmk. I originally made a few branch, pulled the user repo and submitted changes as another PR. Then I realised that this wasn't necessary. It is possible to pull commits from a user repo (their branch) and merge locally. Pushing to the organisation repo merges the PR via the command-line. I recommend this workflow to merge PRs. This way, only admins are handling branches and merges where things get more difficult with Git.

It should be possible to submit a PR without doing this. I want to lower the barrier for entry as much as possible. Creating a fork or pulling to a local repo creates a different branch specific that machine/account already. It should be possible to merge. I am willing to merge PRs and resolve conflicts where needed. However it would help if member submitting pull requests would pull from the organisation repo and merge changes before pushing to their account and submitting a pull request. They should also be pulling from the organisation repo regularly to ensure that they're working on the latest version (to reduce conflicts with things others have already changed).

**Comment by rikutakei** [comment]
Holy yeezus, there's a lot to cover :grimacing: but here we go:

---

> I don't recommend people who are inexperienced with git to be handling branches and rebase. 

Yea, I agree that it'll be confusing for beginners to learn branching, but I would argue that `rebase` is a lot easier and cleaner than merging everything when you're pulling. Although I haven't encountered conflicts with `rebase` (yet), I'm guessing it would be a **lot** more straightforward to fix conflicts as it arises via `rebase`, than to fix all the conflicts in one go with merging *every* changes since you last pulled.

> If we push rebased commits to a remote repo, it could be a problem for someone else.

No, this shouldn't be a problem. The idea is that people will do a `pull --rebase` from the main repo **ONLY when their branch is ready for a PR** (and this should only be done (ideally) once) - any subsequent changes for that PR will not require pulling from the main repo, as the PR should be self-contained to fix a specific problem or adding functionality. Once the PR is successfully merged to the main remote repo, *then* it might be a problem for others, as the merged changes may be in conflict with their edits (but at least there won't be any confusing merge commits to clutter what has been changed).

> I encountered a similar issue when resolving conflicts for PR 32 from rkkmk. 

It might be harsh to ask a lot from beginners, but in my opinion, this wouldn't have been a big problem if a `git pull` (regardless of rebasing) was made and conflicts resolved **prior** to the PR.

> I originally made a few branch, pulled the user repo and submitted changes as another PR. 

If you want to add changes (other than merging) to the branch/PR yourself, then this approach is correct. However, merging changes, fixing conflicts, and updating the branch to the latest version was not necessary - this could have been done/resolved during the merging of the pull request (i.e. once all the edits were made for the PR) via the method you described below. Also, by resolving conflicts/etc. at the end of a PR, you can tell from the log that these conflicts were associated with that particular PR.

> It is possible to pull commits from a user repo (their branch) and merge locally. Pushing to the organisation repo merges the PR via the command-line. I recommend this workflow to merge PRs. 

This should only be necessary if there is a merge conflict, but it's up to the user's preference, I guess. 

> This way, only admins are handling branches and merges where things get more difficult with Git.

Merging to the main repo is only done by people with write access (i.e. admins) anyway, so it doesn't matter how "difficult" the `merge` is or how complicated the branching/conflicts/etc. are; Admins are expected to fix these problems (via command line, or whatever) - not the translators and/or reviewers. With that said, PR submitter may have to resolve conflicts before submitting PR - in that case, there will always be help on Slack/TopicBox/email.

> It should be possible to submit a PR without doing this. 

Yes, but with potential chaos down the road for the admins.

> I want to lower the barrier for entry as much as possible. Creating a fork or pulling to a local repo creates a different branch specific that machine/account already. 

Yes, agreed, but branching is just one small extra step for them to do.

> it would help if member submitting pull requests would pull from the organisation repo and merge changes before pushing to their account and submitting a pull request. 

And it would be a lot easier for admins to check what changes actually need to be reviewed if there are minimal merge commits.

> They should also be pulling from the organisation repo regularly to ensure that they're working on the latest version (to reduce conflicts with things others have already changed).

I disagree. I don't know how `git` works inside out, but how does it assign merge commit hash? Hypothetically, if everyone decided to `pull` and had to `merge` the changes, there will be N merge commits (unless `git` is smart and uses the same hash for everyone). And what will happen when everyone (with their different merge commit hash) sends a PR separately? There will be a LOT of unnecessary merge commits in the log for the "same" `merge` that other people had to do. And hence my request to use merge commits only for PR, if possible.

---

I would propose `git pull` before you start making changes, switch to a different branch, and a `git pull --rebase` when you're ready for a PR. This way, there will be minimal merge commits and we can concentrate on the changes made by the submitter.

**Comment by TomKellyGenetics** [comment]
I agree that we should include pull from the main repo in the workflow. This will:

- keep the lessons you're working on up to date.

- reduce the number of conflicts to resolve.

I've made the ".mo" files ignored by git. We don't need to track these. I think they're a file generated by the PO editor (kinda like how MS word makes ~ files). We don't need to merge each others versions of these.

I don't have a problem with people using rebase or merge in their own workflows, provided that they don't rebase commits that have already been pushed to a remote branch (i.e., shared with others). How about we _recommend_ to rebase to reduce the number of merge commits but leave it up to their discretion if they don't feel confident with it. To be more accepting, I think it's better to recommend best practices for the workflow.

I'm ambivalent about branches, it's not that difficult but we won't even know if they do it or not on their local repo. Again this could be included in recommendations.

**Comment by rikutakei** [comment]
Yep, okay, recommendation sounds like a better option for rebasing/merging, etc.
I'll make it clear to ensure people pull first and solve conflicts before pushing to make things easier for us.

We can leave out the branching for now, and chuck it in the recommendations too.

**Review by joelnitta — APPROVED** [review]

**Review by TomKellyGenetics — APPROVED** [review]
@rikutakei I'm okay for you to merge this if you're happy with the current version

**Pull request merged**
